scripts/eqtl_hail_batch/conditional_analysis.py

```diff
@@ -0,0 +1,459 @@
+    # data (this varies by cell type)
+    samples_to_keep = set(residual_df.sampleid)
+    set_to_keep = hl.literal(samples_to_keep)
+    mt = mt.filter_cols(set_to_keep.contains(mt['onek1k_id']))
+    # Do this only on SNPs contained within gene_snp_df to save on
+    # computational time
+    snps_to_keep = set(gene_snp_test_df.snpid)
+    sorted_snps = sorted(snps_to_keep)
+    sorted_snp_positions = list(map(lambda x: x.split(':')[:2][1], sorted_snps))
+    sorted_snp_positions = [int(i) for i in sorted_snp_positions]
+    # get first and last positions, with 1 added to last position (to make it inclusive)
+    chromosome = gene_snp_test_df.snpid[0].split(':')[:1][0]
+    first_and_last_snp = chromosome + ':' + str(sorted_snp_positions[0]) + '-' + str(sorted_snp_positions[-1]+1)
+    mt = hl.filter_intervals(mt, [hl.parse_locus_interval(first_and_last_snp, reference_genome='GRCh38')])
+    t = mt.entries()
+    t = t.annotate(n_alt_alleles=t.GT.n_alt_alleles())
+    t = t.key_by(contig=t.locus.contig, position=t.locus.position)
+    t = t.select(t.alleles, t.n_alt_alleles, sampleid=t.onek1k_id)
+    t = t.annotate(snpid=hl.str(':').join(list(map(hl.str, [t.contig, t.position, t.alleles[0], t.alleles[1]]))))
+    # Further reduce the table by only selecting SNPs needed
+    set_to_keep = hl.literal(snps_to_keep)
+    t = t.filter(set_to_keep.contains(t['snpid']))
+    # only keep SNPs where all samples have an alt_allele value
+    snps_to_remove = set(t.filter(hl.is_missing(t.n_alt_alleles)).snpid.collect())
+    if len(snps_to_remove) > 0:
+        t = t.filter(~hl.literal(snps_to_remove).contains(t.snpid))
+
+    genotype_df = t.to_pandas(flatten=True)
+    genotype_df.rename({'onek1k_id': 'sampleid'}, axis=1, inplace=True)
+
+    return genotype_df
+
+
+def calculate_residual_df(residual_df, significant_snps_df, filtered_mt_path):
+    """calculate residuals for gene list"""
+    
+    # make sure 'gene_symbol' is the first column
+    # otherwise, error thrown when using reset_index
+    cols = list(significant_snps_df)
+    cols.insert(0, cols.pop(cols.index('gene_symbol')))
+    significant_snps_df = significant_snps_df.loc[:, cols]
+
+    # Identify the top eSNP for each eGene 
+    esnp1 = (
+        significant_snps_df.sort_values(['gene_symbol', 'fdr'], ascending=True)
+        .groupby('gene_symbol')
+        .first()
+        .reset_index()
+    )
+
+    # Subset residuals for the genes to be tested
+    gene_ids = esnp1['gene_symbol'][esnp1['gene_symbol'].isin(residual_df.columns)].to_list()
+    # save sampleids before filtering redidual_df
+    sample_ids = residual_df.sampleid
+    residual_df = residual_df[gene_ids]
+    # reassign sample ids
+    residual_df = residual_df.assign(sampleid=sample_ids.to_list())
+
+    # Subset genotype file for the significant SNPs
+    genotype_df = get_genotype_df(
+        filtered_mt_path, residual_df, esnp1
+    )
```

Comment:
I'm somewhat struggling to follow what's happening here - sorry, I think my python is too rusty! But I can't find when the lead SNP is actually being adjusted and how..

So is `esnp1` now all significant variants in the locus _minus_ the top SNP (did `.reset_index()` on line 142 drop it?)? And then we're now fetching the genotypes for all those variants in `esnp1`?

But if so, later on line 164 how are we managing to retrieve the ID of the top SNP from `esnp1` and `genotype_df`?

Reply:
esnp1 is the most significant SNP (reset index simply moves the gene_name, which would otherwise be the index, to within the df). get_genotype_df() then gets the genotypes for all samples, exclusively for that SNP. A linear regression model is then fit to the data (residual expression values for each sample x genotype), and the residuals of that are taken

Reply:
Hmm, ok, I think I see - it had confused me because the original script [pulled all genotypes](https://github.com/powellgenomicslab/onek1k_phase1/blob/main/single_cell_cis_eQTL_mapping/round2.run_spearman_rank_test.R#L103), but I see now that that is unnecessary. So am I interpreting it right that the process is basically this?

**Base analysis (Round 1)**
- _Input: all SNP genotypes, full scRNA-seq expression data_
- **Step 1**: fit a model to expression data to account for experimental covariates (sex, age, PCAs 1-4, Peer Factors 1 & 2), take residuals (= adjusted expression)
- **Step 2**: test association of all SNPs to adjusted expression
- _Output: set of significantly associated SNPs, expression data adjusted for covariates_

**Conditional analysis (Round 2 +)**
- _Input: significant SNP genotypes from previous round, expression data adjusted for experimental covariates (or in round 3+, adjusted for experimental covariates and previous round lead SNPs)_
- **Step 1**: fit a model to adjusted expression data to account for genotype of most significant SNP in locus, take residuals (= SNP & covariate adjusted expression)
- **Step 2**: test association of significant SNPs from previous round against SNP & covariate adjusted expression data 
- _Output: set of significantly associated SNPs, expression data adjusted for lead SNP & covariates_

Reply:
yes, that's exactly right!

Reply:
Thanks! Not a blocker on this PR, but if you have time to update the README to explain the process then I at least would find that helpful when coming back to this months later :)

Reply:
Fantastic, I can add that in there.